libs/datamodel/core/src/common/preview_features.rs

```diff
@@ -1,57 +1,155 @@
+            pub fn parse_opt(s: &str) -> Option<Self> {
+                paste! {
+                    let parsed = match s.to_lowercase().as_str() {
+                        $(
+                            stringify!([<$variant:lower>]) => Self::$variant,
+                        )*
+                        _ => return None,
+                    };
+                }
+
+                Some(parsed)
+            }
+        }
+
+        impl ToString for PreviewFeature {
+            fn to_string(&self) -> String {
+                match self {
+                    $(
+                        Self::$variant => decapitalize(stringify!($variant)),
+                    )*
+                }
+            }
+        }
+    };
+}
+
+// (Usually) Append-only list of features.
+features!(
+    ConnectOrCreate,
+    TransactionApi,
+    NativeTypes,
+    GroupBy,
+    CreateMany,
+    AtomicNumberOperations,
+    AggregateApi,
+    Middlewares,
+    Distinct,
+    UncheckedScalarInputs,
+    SqlServer,
+    MongoDb,
+    OrderByRelation,
+    NApi,
+    SelectRelationCount,
+    OrderByAggregateGroup
+);
+
+// Mapping of which active, deprecated and hidden
+// features are valid in which place in the datamodel.
+lazy_static! {
+    /// Generator preview features
+    pub static ref GENERATOR: FeatureMap = {
+        FeatureMap::default().with_active(vec![
+            SqlServer,
+            OrderByRelation,
+            NApi,
+            SelectRelationCount,
+            OrderByAggregateGroup
+        ]).with_hidden(vec![
+            MongoDb
+        ]).with_deprecated(vec![
+            AtomicNumberOperations,
+            AggregateApi,
+            Middlewares,
+            NativeTypes,
+            Distinct,
+            ConnectOrCreate,
+            TransactionApi,
+            UncheckedScalarInputs,
+            GroupBy,
+            CreateMany
+        ])
+    };
+
+    /// Datasource preview features.
+    pub static ref DATASOURCE: FeatureMap = {
+        FeatureMap::default()
+    };
+}
+
+#[derive(Debug, Default)]
+pub struct FeatureMap {
+    /// Valid, visible features.
+    active: Vec<PreviewFeature>,
+
+    /// Deprecated features.
+    deprecated: Vec<PreviewFeature>,
+
+    /// Hidden preview features are valid features, but are not propagated into the tooling
+    /// (as autocomplete or similar) or into error messages (eg. showing a list of valid features).
+    hidden: Vec<PreviewFeature>,
+}
+
+impl FeatureMap {
+    pub fn active_features(&self) -> &[PreviewFeature] {
+        &self.active
+    }
+
+    pub fn hidden_features(&self) -> &[PreviewFeature] {
+        &self.hidden
+    }
+
+    pub fn with_active(mut self, active: Vec<PreviewFeature>) -> Self {
+        self.active = active;
+        self
+    }
+
+    pub fn with_hidden(mut self, hidden: Vec<PreviewFeature>) -> Self {
+        self.hidden = hidden;
+        self
+    }
+
+    pub fn with_deprecated(mut self, deprecated: Vec<PreviewFeature>) -> Self {
```

Comment:
nitpick: do these three methods need to be public?

Reply:
Probably not.